pkg/cni/plugin.go

```diff
@@ -0,0 +1,259 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cni
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"runtime"
+	"strings"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+const (
+	// IPv4InterfaceArpProxySysctlTemplate allows proxy_arp on a given interface
+	IPv4InterfaceArpProxySysctlTemplate = "net.ipv4.conf.%s.proxy_arp"
+)
+
+// A NetConf structure represents a Multus network attachment definition configuration
+type NetConf struct {
+	types.NetConf
+	DeviceID string `json:"deviceID"`
+	MTU      int    `json:"mtu,omitempty"`
+}
+
+// EnvArgs structure represents inputs sent from each VMI via environment variables
+type EnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(bytes []byte) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(bytes, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func getEnvArgs(envArgsString string) (EnvArgs, error) {
+	if envArgsString != "" {
+		e := EnvArgs{}
+		err := types.LoadArgs(envArgsString, &e)
+		if err != nil {
+			return EnvArgs{}, err
+		}
+		return e, nil
+	}
+	return EnvArgs{}, nil
+}
+
+func getMTUByName(ifName string) (int, error) {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return 0, err
+	}
+	return link.Attrs().MTU, nil
+}
+
+func configureArp(macvtapLink netlink.Link, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		// For sysctl, dots are replaced with forward slashes
+		ifaceNameAllowingDots := strings.Replace(macvtapLink.Attrs().Name, ".", "/", -1)
+
+		// TODO: duplicate following lines for ipv6 support, when it will be added in other places
+		ipv4SysctlValueName := fmt.Sprintf(IPv4InterfaceArpProxySysctlTemplate, ifaceNameAllowingDots)
+		if _, err := sysctl.Sysctl(ipv4SysctlValueName, "1"); err != nil {
+			// remove the newly added link and ignore errors, because we already are in a failed state
+			netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to set proxy_arp on newly added interface %q: %v", macvtapLink.Attrs().Name, err)
+		}
+		return nil
+	})
+	return err
+}
+
+func configureMacvtap(conf *NetConf, macAddr net.HardwareAddr, ifName string, netns ns.NetNS) (*current.Interface, error) {
```

Comment:
I'll take a better look into this suggestion. Expect an update to the PR in the following day(s).

Reply:
Updated version takes into account (and accepts) most of this suggestion.

I just don't see the benefit of not having `configureArp` isolated into it's own function.

Reply:
> I just don't see the benefit of not having `configureArp` isolated into it's own function.

Fair enough.

> Updated version takes into account (and accepts) most of this suggestion.

This might be a cosmetic thing and I might be missing something, but you end up passing the namespace around in multiple functions having the ns.Do block in each. What is preventing you to invert the pattern and move all those funcion calls within a single ns.Do block?
So instead of
```
CmdAdd() {
  ConfigureMacvtap(..., ns) {
    ns.Do {
      ...
    }
    renameMacvtapIface(..., ns) {
      ns.Do() {
        ...
      }
    }
  }
  LinkSetUp(..., ns) {
    ns.Do {
      ...
    }
  }
}
```
make it like
```
CmdAdd() {
  ns.Do {
    ConfigureMacvtap(...) {
      ...
      renameMacvtapIface(...) {
        ...
      }
    }
    LinkSetUp(...) {
      ...
    }
  }
}
```

Reply:
> > I just don't see the benefit of not having `configureArp` isolated into it's own function.
> 
> Fair enough.
> 
> > Updated version takes into account (and accepts) most of this suggestion.
> 
> This might be a cosmetic thing and I might be missing something, but you end up passing the namespace around in multiple functions having the ns.Do block in each. What is preventing you to invert the pattern and move all those funcion calls within a single ns.Do block?
> So instead of
> 
> ```
> CmdAdd() {
>   ConfigureMacvtap(..., ns) {
>     ns.Do {
>       ...
>     }
>     renameMacvtapIface(..., ns) {
>       ns.Do() {
>         ...
>       }
>     }
>   }
>   LinkSetUp(..., ns) {
>     ns.Do {
>       ...
>     }
>   }
> }
> ```
> 
> make it like
> 
> ```
> CmdAdd() {
>   ns.Do {
>     ConfigureMacvtap(...) {
>       ...
>       renameMacvtapIface(...) {
>         ...
>       }
>     }
>     LinkSetUp(...) {
>       ...
>     }
>   }
> }
> ```

Reduced the usage of `netns.Do(...)` to a single usage, if only using a different pattern.